collect_app/src/test/java/org/odk/collect/android/widgets/RangePickerDecimalWidgetTest.java

```diff
@@ -0,0 +1,128 @@
+package org.odk.collect.android.widgets;
+
+import android.view.View;
+
+import org.javarosa.core.model.RangeQuestion;
+import org.javarosa.core.model.data.StringData;
+import org.javarosa.form.api.FormEntryPrompt;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.R;
+import org.odk.collect.android.formentry.questions.QuestionDetails;
+import org.odk.collect.android.listeners.WidgetValueChangedListener;
+import org.robolectric.RobolectricTestRunner;
+import java.math.BigDecimal;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.mockValueChangedListener;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithQuestionDefAndAnswer;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnlyAndQuestionDef;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;
+
+@RunWith(RobolectricTestRunner.class)
+public class RangePickerDecimalWidgetTest {
+
+    private RangeQuestion rangeQuestion;
+
+    @Before
+    public void setup() {
+        rangeQuestion = mock(RangeQuestion.class);
+
+        when(rangeQuestion.getRangeStart()).thenReturn(new BigDecimal("1.5"));
+        when(rangeQuestion.getRangeEnd()).thenReturn(new BigDecimal("5.5"));
+        when(rangeQuestion.getRangeStep()).thenReturn(new BigDecimal("0.5"));
+    }
+
+    @Test
+    public void whenRangeEndIsGreaterThanRangeStart_widgetsSetsCorrectValuesForNumberPicker() {
+        RangePickerDecimalWidget widget = createWidget(promptWithQuestionDefAndAnswer(rangeQuestion, null));
+        int index = 0;
+
+        for (double i = 5.5F; i >= 1.5F; i -= 0.5F, index++) {
+            assertThat(widget.displayedValuesForNumberPicker[index], equalTo(String.valueOf(i)));
```

Comment:
I'm not a fan of testing this field here as there is no reason for it to be public. Really the test should click on the button and test that the dialog receives the right values. I know you're reworking how all that functions though so I think we go with this for the moment!

As a general tip for tests like this I would avoid the `for` loop and just write out what I expect the values to be. That way it's clearer to the reader of the test. Tests serve as documentation as much as they serve as checks so it's good for them to be very clear what the behaviour they expect is.